metabase/metabase
# Let users create a new collection when saving or moving things

Closes #4020
Closes #29774

### Description

Adds a new option for creating a collection “on the go” when inside other modals that have a collection picker.

* User can’t click `New collection` until the rest of the form is valid.
* When user clicks `New collection`, the current modal is replaced by the New Collection modal.
* In New Collection modal, when user clicks `Close` they go back to the previous modal.
* In New Collection modal, when user clicks `Create` the collection is created, then the previous modal form is submitted to complete the parent action.

Here are the affected components and their new flows.  The new buttons are shown in red:

![CleanShot 2023-06-21 at 15 32 14@2x](https://github.com/metabase/metabase/assets/116838/83539122-b886-4348-97a3-2688d1e7b788)

### How to verify

Try each of the flows in the screenshot above:
* Move an item → to a collection created on the go.
* Create a dashboard → in a collection created on the go.
* Save a question → in a collection created on the go.

Make sure the collection exists with the new item added to it.

### Changes

* Some modals were modernized to functional components to simplify the additions needed by this feature.
* A new Button flag `light` was added for the `+ New collection` button style seen in the figma doc.

### Notes for Code Review

* 👀Looking at the separate commit diffs will be easier to review. 
* To handle the “modal in a modal” situation, each parent modal has a `creatingCollection` boolean state flag to temporarily replace its own modal with the New Collection modal.
  * (This is the same pattern used for creating dashboards on the go.)
* When canceling the New Collection modal, we restore the original modal’s field values— which are staged in a state variable when the user clicks “new collection”.
* I implemented the same pattern in subtly different ways in three separate modal components.  Not yet sure how to unify them into a single, reusable thing.

### Demo


### Checklist

- [x] Restore the previously selected collection when canceling New Collection from the Move modal.
- [x] “New Collection” should be created relative to the selected collection in previous modal
